README
Hessian-Client

A Hessian Client for perl.
Currently, Hessian 1.0 is supported;
Hessian 2.0 is also tested to work, although specification is still "draft";

With little dependencies, it should be easy to install;
Also you may want to write a wrapper to suit your need;

Also a hessian service is needed if you want to test this module;


INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Hessian::Client

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Hessian-Client

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Hessian-Client

    CPAN Ratings
        http://cpanratings.perl.org/d/Hessian-Client

    Search CPAN
        http://search.cpan.org/dist/Hessian-Client


COPYRIGHT AND LICENCE

Copyright (C) 2009 du ling

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Revision history for Hessian-Client

0.01    Thu May 21 21:55:03 UTC 2009
        First version, released on an unsuspecting world.
0.02    Thu May 21 22:33:37 UTC 2009
        fix reading of compact binary, in hessian 2.0



t/lib/Communication/TestServlet2.pm
package  Communication::TestServlet2;

use strict;
use warnings;

use base 'Communication';

use Test::More;
use Test::Deep;
use Test::Exception;

use Hessian::Client;
use YAML;
use DateTime;

my $test_service = 'http://hessian.caucho.com/test/test2';

sub prep01_check_webservice : Test(startup) {    #{{{
    my $self   = shift;
    my $client = Hessian::Client->new(
        {
            version => 2,
            service => $test_service
        }
    );
    eval {

        my $result = $client->methodNull();
    };
    if ( my $e = $@ ) {
        $self->SKIP_ALL("Problem connecting to test service.");
    }

}    #}}}

sub test_reply_int_0 : Test(1) {    #{{{
    my $self   = shift;
    my $client = Hessian::Client->new(
        {
            version => 2,
            service => $test_service
        }
    );
    my $result = $client->replyInt_0();
    is( $result->{reply_data}, 0 );
}    #}}}

sub test_reply_int_47 : Test(1) {    #{{{
    my $self   = shift;
    my $client = Hessian::Client->new(
        {
            version => 2,
            service => $test_service
        }
    );
    my $result = $client->replyInt_47();
    is( $result->{reply_data}, 47 );
}    #}}}

sub test_reply_int_mx800 : Test(1) {    #{{{
    my $self   = shift;
    my $client = Hessian::Client->new(
        {
            version => 2,
            service => $test_service
        }
    );
    my $function = "replyInt_m0x800";
    my $result   = $client->$function();
    is( $result->{reply_data}, -0x800 );
}    #}}}

sub test_reply_long_mOx80000000 : Test(1) {    #{{{
    my $self   = shift;
    local $TODO = "Platform dependent, may not pass.";
    my $client = Hessian::Client->new(
        {
            version => 2,
            service => $test_service
        }
    );
    my $function = "replyLong_m0x80000000";
    my $result   = $client->$function();
    is( $result->{reply_data}, -0x80000000 );
}    #}}}

sub test_reply_long_mOx80000001 : Test(1) {    #{{{
    my $self   = shift;
    my $client = Hessian::Client->new(
        {
            version => 2,
            service => $test_service
        }
    );
    my $function = "replyLong_m0x80000001";
    my $result   = $client->$function();
    is( $result->{reply_data}, -0x80000001 );
}    #}}}

sub test_reply_long_Ox10 : Test(1) {    #{{{
    my $self   = shift;
    my $client = Hessian::Client->new(
        {
            version => 2,
            service => $test_service
        }
    );
    my $function = "replyLong_0x10";
    my $result   = $client->$function();
    is( $result->{reply_data}, 0x10 );
}    #}}}

sub test_reply_double_0_0 : Test(1) {    #{{{
    my $self   = shift;
    my $client = Hessian::Client->new(
        {
            version => 2,
            service => $test_service
        }
    );
    my $function = "replyDouble_0_0";
    my $result   = $client->$function();
    is( $result->{reply_data}, 0.0 );
}    #}}}

sub test_reply_double_m0_001 : Test(1) {    #{{{
    my $self = shift;
    local $TODO = "Haven't worked out 32 bit doubles yet.";
    my $client = Hessian::Client->new(
        {
            version => 2,
            service => $test_service
        }
    );
    my $function = "replyDouble_m0_001";
    my $result   = $client->$function();
    is( $result->{reply_data}, -0.001 );
}    #}}}

sub test_reply_double_127_0 : Test(1) {    #{{{
    my $self   = shift;
    my $client = Hessian::Client->new(
        {
            version => 2,
            service => $test_service
        }
    );
    my $function = "replyDouble_127_0";
    my $result   = $client->$function();
    is( $result->{reply_data}, 127 );
}    #}}}

sub test_reply_double_3_14159 : Test(1) {    #{{{
    my $self   = shift;
    my $client = Hessian::Client->new(
        {
            version => 2,
            service => $test_service
        }
    );
    my $function = "replyDouble_3_14159";
    my $result   = $client->$function();
    is( $result->{reply_data}, 3.14159 );
}    #}}}

sub test_reply_int_m17 : Test(1) {    #{{{
    my $self   = shift;
    my $client = Hessian::Client->new(
        {
            version => 2,
            service => $test_service
        }
    );
    my $result = $client->replyInt_m17();
    is( $result->{reply_data}, -17 );
}    #}}}

sub reply_object_16 : Test(1) {    #{{{
    my $self           = shift;
    my $hessian_client = Hessian::Client->new(
        {
            version => 2,
            service => $test_service
        }
    );
    my $result = $hessian_client->replyObject_16();
    cmp_deeply(
        $result,
        { hessian_version => '2.0', reply_data => array_each( ignore() ) },
        "Received expected datastructure."
    );
}    #}}}

sub reply_object_2b : Test(1) {    #{{{
    my $self   = shift;
    my $client = get_client();
    my $result = $client->replyObject_2b();
    my $array  = $result->{reply_data};
    cmp_deeply( $array->[0], $array->[1],
        "Received a two element array of the same object." );
}    #}}}

sub reply_date_0 : Test(1) {    #{{{
    my $self = shift;
    my $date = DateTime->new(
        year      => 1970,
        month     => 1,
        day       => 1,
        minute    => 0,
        hour      => 0,
        time_zone => 'UTC'
    );
    my $client = Hessian::Client->new(
        {
            version => 2,
            service => $test_service
        }
    );
    my $result = $client->replyDate_0();
    is( DateTime->compare( $result->{reply_data}, $date ), 0 );

}    #}}}

sub reply_date_1 : Test(1) {    #{{{
    my $self = shift;
    local $TODO = "2hr Discrepency in time calculation.";
    my $date = DateTime->new(
        year       => 1998,
        month      => 8,
        day        => 5,
        second     => 31,
        minute     => 51,
        nanosecond => 0,
        hour       => 7,
        time_zone  => 'GMT'
    );
    my $client = Hessian::Client->new(
        {
            version => 2,
            service => $test_service
        }
    );
    my $result      = $client->replyDate_1();
    my $result_date = $result->{reply_data};
    is( DateTime->compare( $result_date, $date ), 0 );

}    #}}}

sub reply_date_2 : Test(1) {    #{{{
    my $self = shift;
    local $TODO = "2hr Discrepency in time calculation.";
    my $date = DateTime->new(
        year      => 1998,
        month     => 8,
        day       => 5,
        minute    => 51,
        hour      => 7,
        time_zone => 'GMT'
    );
    my $client      = get_client();
    my $result      = $client->replyDate_2();
    my $result_date = $result->{reply_data};
    is( DateTime->compare( $result_date, $date ), 0 );

}    #}}}

sub reply_untyped_fixed_list_7 : Test(1) {    #{{{
    my $self   = shift;
    my $client = get_client();
    my $result;
    lives_ok {
        $result = $client->replyUntypedFixedList_7();
    }
    "No problems communicating with service.";

}    #}}}

sub get_client {    #{{{

    my $client = Hessian::Client->new(
        {
            version => 2,
            service => $test_service
        }
    );
    return $client;
}    #}}}

"one, but we're not the same";

__END__


=head1 NAME

Communication::TestServlet2 - Test communication with a test service that runs
version 2

=head1 SYNOPSIS

=head1 DESCRIPTION

=head1 INTERFACE




lib/Hessian/Type.pm
package Hessian::Type;

use warnings;
use strict;

=head1 NAME

Hessian::Type - The great new Hessian::Type!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';


=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use Hessian::Type;

    my $foo = Hessian::Null->new();
    
    my $bar = Hessian::List->new(
        type => 'java.lang.String',
        len  => 1,
        data => [
                'hello',
                'world',
                ]
    );
    ...

=head1 FUNCTIONS

=head2 new

=cut

sub new {
    my($class,@params) = @_;
    my $self = {};
    %$self = (@params);
    bless $self, $class;
    return $self;
}

package Hessian::Null;
use vars qw(@ISA);
@ISA = qw(Hessian::Type);

package Hessian::True;
use vars qw(@ISA);
@ISA = qw(Hessian::Type);

package Hessian::False;
use vars qw(@ISA);
@ISA = qw(Hessian::Type);

package Hessian::Binary;
use vars qw(@ISA);
@ISA = qw(Hessian::Type);

package Hessian::XML;
use vars qw(@ISA);
@ISA = qw(Hessian::Type);

package Hessian::Double;
use vars qw(@ISA);
@ISA = qw(Hessian::Type);

package Hessian::List;
use vars qw(@ISA);
@ISA = qw(Hessian::Type);

package Hessian::Map;
use vars qw(@ISA);
@ISA = qw(Hessian::Type);

package Hessian::Remote;
use vars qw(@ISA);
@ISA = qw(Hessian::Type);

package Hessian::Fault;
use vars qw(@ISA);
@ISA = qw(Hessian::Type);

#Hessian 2.0
package Hessian::Class;
use vars qw(@ISA);
@ISA = qw(Hessian::Type);

package Hessian::Object;
use vars qw(@ISA);
@ISA = qw(Hessian::Type);
=head1 AUTHOR

du ling, C<< <ling.du at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-hessian-type at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Hessian-Client>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Hessian::Client


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Hessian-Client>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Hessian-Client>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Hessian-Client>

=item * Search CPAN

L<http://search.cpan.org/dist/Hessian-Client>

=back


=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2009 du ling, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut

1; # End of Hessian::Type


lib/Hessian/Client.pm
package Hessian::Client;

use warnings;
use strict;

use LWP::UserAgent;
use HTTP::Headers;
use List::Util qw/first/;

use Hessian::Translator;
use Hessian::Translator::V1;
use Hessian::Translator::V2;
use Hessian::Type;

=head1 NAME

Hessian::Client - Hessian 1.0 client in perl

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';


=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use Hessian::Client;

    my $foo = Hessian::Client->new(
        URL => 'http://some.hessian.service/...',
        version => 1,
        auth    => ['user','pass'], # if needed
    );
    my $arg1 = Hessian::True->new();
    my $arg2 = 500;
    my $arg3 = Unicode::String->new('hello world!');
    my $bar = $foo->call('method',$arg1,$arg2,$arg3);
    die $foo->{_error} unless defined $bar;
    ...

=head1 FUNCTIONS

=head2 new

=cut

sub new {
    my($class,@params) = @_;
    my $self = {@params};
    bless $self, $class;

    if($self->{URL}){
        $self->{URI} = URI->new($self->{URL});
        unless($self->{URI}->scheme() eq 'http'){
            return $self->raise_error('URI scheme not http: ', $self->{URI});
        }
    }else{
        return $self->raise_error('URL required in constructor');
    }
    return $self;
}


=head2 call

=cut

sub call {
    my ($self, $method, @params) = @_;

    # create UserAgent
    my $ua = LWP::UserAgent->new;
    $ua->agent("Perl Hessian1.0 Client/$VERSION");

    my $header = HTTP::Headers->new();

    if('ARRAY' eq ref $self->{auth}){
        $header->authorization;
        $header->authorization_basic($self->{auth}->[0],$self->{auth}->[1]);
    }

    my $req= HTTP::Request->new(POST => $self->{URI}, $header);

    if($self->{version} && $self->{version} == 2){
        $self->{translator} = Hessian::Translator::V2->new(debug=>$self->{debug},mangle=>$self->{mangle});
    }else{
        $self->{translator} = Hessian::Translator::V1->new(debug=>$self->{debug},mangle=>$self->{mangle});
    }

    my $content;
    my $tr= $self->{translator};
    eval{
        $content = $tr->writeCall($method, @params);
        1;
    }or return $self->raise_error($@);

    $req->content($content);
    $tr->debug($tr->bin_debug($content)) if $self->{debug};
    my $resp= $ua->request($req);

    $tr->debug("Hessian http response: ",$resp->status_line);

    if($resp->is_success){

        $tr->debug($tr->bin_debug($resp->content)) if $self->{debug};

        eval{
            $self->{reply} = $tr->parseResponse($resp->content);
            1;
        }or return $self->raise_error($@);
        my $f = first {'Hessian::Fault' eq ref $_} @{$self->{reply}};
        return $self->raise_error($tr->dump_fault($f)) if $f;
        my $r = first {ref($_) !~ /^Hessian::(Header|Class)/} @{$self->{reply}};
        return $r;
    }else{

        $tr->debug($resp->content) if $self->{debug};
        return $self->raise_error('Hessian http response unsuccessful: ', $resp->status_line, $resp->error_as_HTML);
    }
}

=head2 raise_error

=cut
sub raise_error { # return undef
    my $self = shift;
    $self->{_error} = join "\n", @_;
    return undef;
}


=head1 Hessian 1.0 Data types mapping

=over 2

=item null

Hessian::Null->new();

=item true

Hessian::True->new();

=item false

Hessian::False->new();

=item date

DateTime (cpan)

=item int

normal perl integer

=item long

Math::BigInt

=item double

Hessian::Double->new(data=>20.09);

=item binary

Hessian::Binary->new(data=>'abcd');

=item string

Unicode::String->new('abcd');

=item xml

Hessian::XML->new(data=> Unicode::String->new('abce'));

=item remote

Hessian::Remote->new(type=> 'bean.x', url => 'http://bean.home');

=item list

  Hessian::List->new(
    type => 'int[',  # optional
    len  => 3,       # optional
    data => [0,1,2],
  );

=item map

  my $joe = Hessian::Map->new(
    type => 'com.caucho.hessian.test.Person',
    data => [
        Unicode::String->new('name') => Unicode::String->new('joe')
    ]
  );

  Hessian::Map->new(
    type => 'com.caucho.hessian.test.Horse',
    data => [
        age => 4,
        sex => Unicode::String->new('m'), 
        owner => $joe
    ],
  );

=back


=head1 AUTHOR

du ling, C<< <ling.du at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-hessian-client at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Hessian-Client>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Hessian::Client


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Hessian-Client>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Hessian-Client>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Hessian-Client>

=item * Search CPAN

L<http://search.cpan.org/dist/Hessian-Client>

=back


=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2009 du ling, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut

1; # End of Hessian::Client


lib/Hessian/Deserializer/Numeric.pm
package  Hessian::Deserializer::Numeric;

use Moose::Role;

use Math::Int64 qw/int64_to_number int64_to_net int64 net_to_int64/;
use Math::BigInt lib   => 'GMP';
use Math::BigFloat lib => 'GMP';
use Hessian::Exception;
use POSIX qw/floor ceil/;
use Switch;

sub read_boolean {    #{{{
    my ( $self, $hessian_value ) = @_;
    return
        $hessian_value =~ /T/ ? 1
      : $hessian_value =~ /F/ ? 0
      :                         die "Not an acceptable boolean value";
}    #}}}

sub read_integer {    #{{{
    my ( $self, $hessian_data ) = @_;
    ( my $raw_octets = $hessian_data ) =~ s/^I(.*)/$1/;
    my @chars = unpack 'C*', $raw_octets;
    my $octet_count = scalar @chars;
    my $result =
        $octet_count == 1 ? _read_single_octet( $chars[0], 0x90 )
      : $octet_count == 2 ? _read_double_octet( \@chars, 0xc8 )
      : $octet_count == 3 ? _read_triple_octet( \@chars, 0xd4 )
      :                    _read_quadruple_octet( \@chars );
    return $result;
}    #}}}

sub read_long {    #{{{
    my ( $self, $hessian_data ) = @_;
    ( my $raw_octets = $hessian_data ) =~ s/^(?:L|\x77)(.*)/$1/;
    my @chars       = unpack 'C*', $raw_octets;
    my $array_size  = scalar @chars;
    my $octet_count = scalar @chars;
    my $result =
        $octet_count == 1 ? _read_single_octet( $chars[0], 0xe0 )
      : $octet_count == 2 ? _read_double_octet( \@chars, 0xf8 )
      : $octet_count == 3 ? _read_triple_octet( \@chars, 0x3c )
      : $octet_count == 4 ? Implementation::X->throw(
        error => "32 bit longs
      not currently supported."
      )
      : _read_full_long($raw_octets);    #\@chars );
                                         #_read_quadruple_octet( \@chars,
    return $result;
}    #}}}

sub read_double {    #{{{
    my ( $self, $octet ) = @_;
    my $double_value =
        $octet =~ /\x{5b}/                    ? 0.0
      : $octet =~ /\x{5c}/                    ? 1.0
      : $octet =~ /(?: \x{5d} | \x{5e} ) .*/x ? _read_compact_double($octet)
      : $octet =~ /\x5f/                      ? Implementation::X->throw(
        error => "32 bit doubles not currently supported." )
      : _read_full_double($octet);

    #_read_quadruple_octet_double($octet)
    return $double_value;
}    #}}}

sub _read_single_octet {    #{{{
    my ( $octet, $octet_shift ) = @_;
    my $integer = $octet - $octet_shift;
    return $integer;
}    #}}}

sub _read_double_octet {    #{{{
    my ( $bytes, $octet_shift ) = @_;
    {
        use integer;
        my $integer = ( ( $bytes->[0] - $octet_shift ) << 8 ) + $bytes->[1];
        return $integer;
    }
}    #}}}

sub _read_triple_octet {    #{{{
    my ( $bytes, $octet_shift ) = @_;
    {
        use integer;
        my $integer =
          ( ( $bytes->[0] - $octet_shift ) << 16 ) +
          ( $bytes->[1] << 8 ) +
          $bytes->[2];
        return $integer;
    }
}    #}}}

sub _read_quadruple_long_octet {    #{{{
    my ( $bytes, $octet_shift ) = @_;
    my $big_int   = Math::BigInt->new();
    my $shift_val = 0;
    my $index     = 0;
    foreach my $byte ( reverse @{$bytes} ) {

        #        $index++;
        my $shift_byte = Math::BigInt->new($byte);

        #        $shift_byte->bsub($octet_shift) if $index == 4;
        $shift_byte->blsft($shift_val);
        $big_int->badd($shift_byte);
        $shift_val += 8;
    }
    return $big_int->bstr();
}    #}}}

sub _read_quadruple_octet {    #{{{
    my $bytes = shift;
    return unpack "l", pack "C*", reverse @{$bytes};
}    #}}}

sub _read_quadruple_float_octet {    #{{{
    my $bytes = shift;
    {
        use integer;

        my $shift_val = 0;
        my $sum;
        foreach my $byte ( reverse @{$bytes} ) {
            $sum += $byte << $shift_val;
            $shift_val += 8;
        }
        return $sum;
    }
}    #}}}

sub _read_full_long {    #{{{
    my $string    = shift;
    my $net_int64 = net_to_int64($string);
    return "$net_int64";
}    #}}}

sub _read_compact_double {    #{{{
    my $compact_octet = shift;
    my @chars = unpack 'c*', $compact_octet;
    shift @chars;
    my $chars_size = scalar @chars;
    my $float      = _read_quadruple_float_octet( \@chars );
    return $float;
}    #}}}

sub _read_quadruple_octet_double {    #{{{
    my $octets = shift;
    $octets =~ s/\x5f//;

    #     my @chars = unpack 'C*', $octets;
    #     my @hex_octets =
    #       map { sprintf "%#02x", $_ } @chars;
    #     my $hex_string = join "" => @hex_octets;

    my $double;

    #     $double = str2float($hex_string);
    return $double;
}    #}}}

sub _read_full_double {    #{{{
    my $double = shift;
    ( my $octets = $double ) =~ s/(?:D ) (.*) /$1/x;
    my @chars = unpack 'C*', $octets;
    my $double_value = unpack 'F', pack 'C*', reverse @chars;
    return $double_value;
}    #}}}

sub read_integer_handle_chunk {    #{{{
    my ( $self, $first_bit ) = @_;
    my $input_handle = $self->input_handle();
    my ( $number, $data );
    switch ($first_bit) {
        case /\x49/ {

            #            read $input_handle, $data, 4;
            $data   = $self->read_from_inputhandle(4);
            $number = $self->read_integer($data);
        }
        case /[\x80-\xbf]/ {
            $number = $self->read_integer($first_bit);
        }
        case /[\xc0-\xcf]/ {

            #            read $input_handle, $data, 1;
            $data   = $self->read_from_inputhandle(1);
            $number = $self->read_integer( $first_bit . $data );
        }
        case /[\xd0-\xd7]/ {

            #            read $input_handle, $data, 2;
            $data   = $self->read_from_inputhandle(2);
            $number = $self->read_integer( $first_bit . $data );
        }

    }
    return $number;

}    #}}}

sub read_long_handle_chunk {    #{{{
    my ( $self, $first_bit ) = @_;
    my $input_handle = $self->input_handle();
    my ( $number, $data );
    switch ($first_bit) {
        case /[\xd8-\xef]/ {
            $number = $self->read_long($first_bit);
        }
        case /[\xf0-\xff]/ {

            #            read $input_handle, $data, 1;
            $data   = $self->read_from_inputhandle(1);
            $number = $self->read_long( $first_bit . $data );
        }
        case /[\x38-\x3f]/ {

            #            read $input_handle, $data, 2;
            $data   = $self->read_from_inputhandle(2);
            $number = $self->read_long( $first_bit . $data );
        }
        case /\x59/ {
            $data   = $self->read_from_inputhandle(4);
            $number = $self->read_long($data);

        }
        case /\x4c/ {

            #            read $input_handle, $data, 8;
            $data   = $self->read_from_inputhandle(8);
            $number = $self->read_long($data);
        }

    }
    return $number;
}    #}}}

sub read_double_handle_chunk {    #{{{
    my ( $self, $first_bit ) = @_;
    my $input_handle = $self->input_handle();
    my ( $number, $data );
    switch ($first_bit) {

        #        case /[\x5b-\x5c]/ { $data = $first_bit; }
        #        case /\x5d/ { read $input_handle, $data, 1; }
        #        case /\x5e/ { read $input_handle, $data, 2; }
        #        case /\x5f/ {
        #            read $input_handle, $data, 4;
        case /[\x5b-\x5c]/ { $data = $first_bit; }
        case /\x5d/ {
            $data = $first_bit . $self->read_from_inputhandle(1);

            #            read $input_handle, $data, 1;
        }
        case /\x5e/ {

            $data = $first_bit . $self->read_from_inputhandle(2);

            #            read $input_handle, $data, 2;
        }
        case /\x5f/ {

            #            read $input_handle, $data, 4;
            $data = $first_bit . $self->read_from_inputhandle(4);
        }
        case /\x44/ {
            $first_bit = "";

            #            read $input_handle, $data, 8;
            $data = $first_bit . $self->read_from_inputhandle(8);
        }

    }
    $number = $self->read_double($data);
    return $number;
}    #}}}

sub read_boolean_handle_chunk {    #{{{
    my ( $self, $first_bit ) = @_;
    return $self->read_boolean($first_bit);

}    #}}}

"one, but we're not the same";

__END__


=head1 NAME

Hessian::Deserializer::Numeric - Deserializer methods for booleans, integers and
floating point numbers.

=head1 VERSION

=head1 SYNOPSIS

These methods are meant for internal use.

=head1 DESCRIPTION

This module provides methods for reading numerical values from the input file
handle.

=head1 INTERFACE

=head2   read_boolean

Read a boolean value.

=head2   read_boolean_handle_chunk

Reads a boolean value from the file handle.


=head2   read_double_handle_chunk

Read a floating point number from the file handle.


=head2   read_integer

Reads an integer.


=head2   read_integer_handle_chunk

Reads an integer.

=head2   read_long

Reads a long numerical value

=head2 read_double

Reads a floating point number.


=head2   read_long_handle_chunk


Reads a long numerical value from the file handle.


lib/Hessian/Translator.pm
package Hessian::Translator;

use warnings;
use strict;

use Math::BigInt;

my $UB32 = Math::BigInt->new(0x7fffffff);
my $LB32 = Math::BigInt->new(-0x80000000);

=head1 NAME

Hessian::Translator - The great new Hessian::Translator!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';



=head1 FUNCTIONS

=head2 new

=cut

sub new {
    my($class,@params) = @_;

    my %h = @params;
    my $self = \%h;

    #these will be used
    $self->{_out} = '';            # call string
    $self->{_in} = '';             # reply string
    $self->{_outRefs} = [];   # ref list for writing call
    $self->{_inRefs} = [];    # ref list for parsing reply

    $self->{_inList} = [];    # reply parsed into perl structure
    $self->{_inHeaders} = []; # reply headers, if used

    # Hessian 2.0 use only
    $self->{_outClassRefs} = []; # class def, Hessian 2.0
    $self->{_inClassRefs} = []; # class def, Hessian 2.0
    $self->{_inTypeRefs} = []; # type, for list/map

    bless $self,$class;
    return $self;
}

=head2 pack_signed_int

=cut

sub pack_signed_int {
    my($self,$x,$n) = @_; #$n = bytes
    $n *= 2; # nybbles = 4-bit
    $x = Math::BigInt->new($x) unless ref($x) eq 'Math::BigInt';
    my $y = $x->as_hex();
    if($x->is_neg()){ #two's complement ...
        my $u = Math::BigInt->new('0x'.'f'x$n);
        $y = $x->bneg()->bxor($u)->binc()->as_hex();
    }
    $y = sprintf "%0${n}s",substr $y,2; #remove leading '0x' and pad with 0
    $y =~ s/([0-9a-f]{2})/pack('H2',$1)/gei;
    return $y;
}

=head2 unpack_int

=cut

sub unpack_int {
    my($self,$x,$i,$len,$signed) = @_; #len in bytes
    my $n = $len * 2;
    $x = substr $x,$i,$len;
    $x =~ s/(.)/sprintf('%02x',ord $1)/ge;
    $x = Math::BigInt->new("0x$x");

    my $y = Math::BigInt->new('0x7'.'f'x($n-1));
    if($signed && $signed !~ /^unsigned$/i && 0 < $x->bcmp($y)){ #signed & negative
        my $u = Math::BigInt->new('0x1'.'0'x$n);
        $x->bsub($u);
    }
    return $x;
}

=head2 is_between

=cut

sub is_between {
    my($self,$x,$l,$u) = @_;
    return $x >= $l && $x <= $u unless ref $x || ref $l || ref $u;
    $x = Math::BigInt->new($x) unless ref $x;
    $l = Math::BigInt->new($l) unless ref $l;
    $u = Math::BigInt->new($u) unless ref $u;
    return 0 <= $x->bcmp($l) && 0 >= $x->bcmp($u) if 'Math::BigInt' eq ref $x;
}

=head2 h2n

=cut

sub h2n { # host 2 network byte order, or n2h
    my($self,$bytes) = @_;
    use Config;
    return $Config{'byteorder'} =~ /^1234/ # little-endian
    ? scalar reverse $bytes
    : $bytes;
}

=head2 bin_debug

=cut

sub bin_debug {
    my($self, $octets) = @_;

    my($cnt,$out) = (1,'');
    for(split//,$octets){
        if( $_ =~ /^[a-zA-Z0-9]/ ){ #newer perl would take /\p{IsGraph}/ ...
            $out .= "$_ ";
        }elsif( $_ =~ /^[\`\~\!\@\#\$\%\^\&\*\(\)\-\_\=\+\[\{\]\}\;\:\'\"\,\<\.\>\/\?]/ ){ #newer perl would take /\p{IsGraph}/ ...
            $out .= "'$_' ";
        }else{
            $out .= sprintf 'x%02x ', ord($_);
        }
        $out .= "\n" if 0 == $cnt++ % 30;
    }
    $out .= "\n";
}

=head2 debug

=cut

sub debug {
    my($self, @msgs) = @_;
    print STDERR "@msgs\n" if $self->{debug};
}

=head2 dump_fault

=cut

sub dump_fault {
    my($self,$f) = @_;
    return unless 'Hessian::Fault' eq ref $f;
    my $stack = join "\n", map{
        join ' ',$_->{__type},$_->{lineNumber},$_->{fileName},"$_->{methodName}()",$_->{declaringClass}
    } grep{'HASH' eq ref $_}@{$f->{detail}->{stackTrace}};
    my $reply = <<EOFAULT;
Hessian::Fault code: $f->{code}
  message: $f->{message}
  detail: $f->{detail}->{__type}: $f->{detail}->{detailMessage}
  $stack
EOFAULT
    print STDERR Data::Dumper::Dumper($reply);
    return $reply;
}

=head2 mangle

=cut

sub mangle {
    my($self,$method,@params) = @_;
    my @pnames = map $self->type_of($_), @params;
    return join '_', $method, @pnames;
}

=head2 type_of

=cut

sub type_of {
    my($self,$x) = @_;
    for(ref $x){
        /^Hessian::Null/        && return 'null';
        /^Hessian::Double/      && return 'double';
        /^Hessian::Long/        && return 'long';
        /^DateTime/             && return 'date';
        /^Unicode::String/      && return 'string';
        /^Hessian::Binary/      && return 'binary';
        /^Hessian::XML/         && return 'xml'; # hessian 1.0 only ?
        /^Hessian::(True|False)/        && return 'boolean';
        /^(Hessian::List|ARRAY)/        && return $x->{type} ? $x->{type} : 'list';
        /^(Hessian::Map|HASH)$/ && return $x->{type} ? $x->{type} : 'map';
        /^Hessian::Remote/      && return 'remote';
        /^REF$/                 && return $self->type_of($$x); # perl ref
        /^$/                    && do {
                                    for($x){
                                        /^[\+-]?\d+$/
                                        && do { # int or long
                                            $x = Math::BigInt->new($x);
                                            return $self->is_between($x,$LB32,$UB32) ? 'int' : 'long';
                                        };
                                        /^[\+-]?(\d+\.|\.\d+)$/ && return 'double';
                                        m!<\?xml version="\d+" .*\?>! && return 'xml';
                                        # default
                                        return 'string';
                                    } # end for $x
                                };
        # DEFAULT
        #die "unkown type_of $x: ", ref $x;
        return 'Object'; # ??
    } # end for ref $x
}

=head1 AUTHOR

du ling, C<< <ling.du at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-hessian-translator at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Hessian-Client>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Hessian::Client


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Hessian-Client>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Hessian-Client>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Hessian-Client>

=item * Search CPAN

L<http://search.cpan.org/dist/Hessian-Client>

=back


=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2009 du ling, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut

1; # End of Hessian::Translator
